### PR TITLE
`plugin` added OpenTelemetry plugin and example

### PR DESCRIPTION
**Pull request by zilto**
Currently, no tests were added.

### CODE REVIEW

File: examples/LLM_Workflows/observability_openllmetry/run.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+if __name__ == "__main__":
+    from traceloop.sdk import Traceloop
+
+    import __main__
+    from hamilton import driver
+    from hamilton.plugins import h_opentelemetry
+
+    Traceloop.init()
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/run.py:26]: Need to explain this code. My reaction is, what if I don't want Traceloop, what do I need to do?

File: examples/LLM_Workflows/observability_openllmetry/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    ```bash
+    python run.py
+    ```
+
+5. Explore results on Traceloop (or your OpenTelemetry destination).
+
+
+## How does it compare to the Hamilton UI
+
+On one hand, OpenTelemetry collects lightweight metadata about dataflow and node executions. It provides a central location for traces when your system has multiple componentes (e.g., server, client, cache, database). On the other hand, the Hamilton UI is tailored to Hamilton and provides visualizations, data lineage, summary statistics, and more utilities to improve your development experience. Given these systems emit lightweight metadata, it's totally reasonable to use both!
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/README.md:38]:
```suggestion
## Should I still use the Hamilton UI?

Short answer is Yes! OpenTelemetry's roots are to collect "telemetry" on the internals of code and external API calls. It is a standard and is most commonly used inside web-services. There isn't any conflict with OpenTelemetry and what is captured in the Hamilton UI. In fact, the Hamilton UI captures a superset of what's possible with OpenTelemetry, and is of course tailored to Hamilton to provide visualizations, data lineage, summary statistics, and more utilities to improve your development experience. 
```

File: examples/LLM_Workflows/observability_openllmetry/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Monitor Hamilton with OpenTelemetry, OpenLLMetry and Traceloop
+
+In this simple example, you'll learn how to use `OpenTelemetryTracer` to trace your Hamilton code, in particular LLM applications.
+
+![Traceloop screenshot](image.png)
+
+[OpenTelemetry](https://opentelemetry.io/) is an open-source cross-language tool that allows to instrument, generate, collect, and export telemetry data (metrics, logs, traces). When building applications and data platforms with several components, OpenTelemetry helps you centralize information about your system allowing you to monitor performance.
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/README.md:7]:
```suggestion
[OpenTelemetry](https://opentelemetry.io/) is an open-source cross-language tool that allows to instrument, generate, collect, and export telemetry data (metrics, logs, traces) in an industry recognized standard way. 
```

File: examples/LLM_Workflows/observability_openllmetry/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Monitor Hamilton with OpenTelemetry, OpenLLMetry and Traceloop
+
+In this simple example, you'll learn how to use `OpenTelemetryTracer` to trace your Hamilton code, in particular LLM applications.
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/README.md:3]:
```suggestion
In this simple example, you'll learn how to use the `OpenTelemetryTracer` to emit traces of your Hamilton code using the OpenTelemetry format, in particular LLM applications.
```

File: examples/LLM_Workflows/observability_openllmetry/run.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    dr = (
+        driver.Builder()
+        .with_modules(__main__)
+        .with_adapters(h_opentelemetry.OpenTelemetryTracer(tracer_name=__name__))
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/run.py:31]: need to explain why this is needed, as opposed to just the `init()` above.

**Review by skrawcz — COMMENTED** [review]
Tightened up wording on what OpenTel is. Otherwise I think you missed committing a file.

File: hamilton/plugins/h_opentelemetry.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+from typing import Any, Collection, List, Optional, Tuple
+
+logger = logging.getLogger(__name__)
+
+
+try:
+    from opentelemetry import context, trace
+    from opentelemetry.sdk.trace import Span
+except ImportError as e:
+    raise ImportError("Failed to import `opentelemetry`") from e
```
Comment (elijahbenizzy) [inline hamilton/plugins/h_opentelemetry.py:13]: We should add instructions with the right installable target.

Comment (elijahbenizzy) [reply hamilton/plugins/h_opentelemetry.py:13]: https://github.com/DAGWorks-Inc/hamilton/blob/f35f6b7fd4abdeabacae49c354b6943925b89dc0/hamilton/plugins/h_ddog.py#L11 for example

File: hamilton/plugins/h_opentelemetry.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        elif tracer_name:
+            self.tracer = trace.get_tracer(tracer_name)
+        else:
+            raise ValueError(
+                "Need to pass `tracer_name` or `tracer`, got: tracer_name=None and tracer=None"
+            )
+
+        self.graph = None
+
+    def run_before_graph_execution(
```
Comment (elijahbenizzy) [inline hamilton/plugins/h_opentelemetry.py:82]: Add task execution as well -- this is where a lot of value comes in (and the distributed notion of tracing shines through). See datadog integration for an example of how: 

https://github.com/DAGWorks-Inc/hamilton/blob/f35f6b7fd4abdeabacae49c354b6943925b89dc0/hamilton/plugins/h_ddog.py#L231

May need something with set_state and setting the current trace, but I think it should work, at least locally with threading.

File: examples/LLM_Workflows/observability_openllmetry/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+5. Explore results on Traceloop (or your OpenTelemetry destination).
+
+### Without Traceloop
+
+For this example to work without Traceloop, you will need to set up your own [OpenTelemetry destination](https://opentelemetry.io/ecosystem/vendors/). We suggest using [Jaeger](https://www.jaegertracing.io/docs/1.47/getting-started/) and included Python code to route telemetry to it in `run.py`.
+
+## Should I still use the Hamilton UI?
+
+Absolutely! OpenTelemetry focsues on collecting telemetry about the interals of code and external API calls. It's a standard amongst web services. There's no conflict between the OpenTelemetry tracer and the tracking for the Hamilton UI. In fact, the Hamilton UI captures a superset of what OpenTelemetry allows, tailored to the Hamilton framework such as:  visualizations, data lineage, summary statistics, and more utilities to improve your development experience. In the not too distant future, the Hamilton UI could ingest OpenTelemetry data 😉 (contributions welcomed!)
```
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/README.md:41]:
```suggestion
Absolutely! OpenTelemetry focsues on collecting telemetry about the internals of code and external API calls. It's a standard amongst web services. There's no conflict between the OpenTelemetry tracer and the tracking for the Hamilton UI. In fact, the Hamilton UI captures a superset of what OpenTelemetry allows, tailored to the Hamilton framework such as:  visualizations, data lineage, summary statistics, and more utilities to improve your development experience. In the not too distant future, the Hamilton UI could ingest OpenTelemetry data 😉 (contributions welcomed!)
```

File: examples/LLM_Workflows/observability_openllmetry/run.py
Comment (skrawcz) [inline examples/LLM_Workflows/observability_openllmetry/run.py:1]: notebook version would be nice too.

**Review by skrawcz — APPROVED** [review]

**Pull request merged**